File: Java/백준/Silver/1992. 쿼드트리/쿼드트리.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    static int N;
    static char[][] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        arr = new char[N][N];
        
        for (int i = 0; i < N; i++) {
            String line = br.readLine();
            arr[i] = line.toCharArray();
        }

        System.out.println(aaa(0, 0, N));
    }

    static boolean check(int x, int y, int size) {
        char start = arr[x][y];
        for (int i = x; i < x + size; i++) {
            for (int j = y; j < y + size; j++) {
                if (arr[i][j] != start) {
                    return false;
                }
            }
        }
        return true;
    }

    static String aaa(int x, int y, int size) {
        if (size == 1) {
            return String.valueOf(arr[x][y]);
        }

        if (check(x, y, size)) {
            return String.valueOf(arr[x][y]);
        }

        int newSize = size / 2;
        String p1 = aaa(x, y, newSize);
        String p2 = aaa(x, y + newSize, newSize);
        String p3 = aaa(x + newSize, y, newSize);
        String p4 = aaa(x + newSize, y + newSize, newSize);

        return "(" + p1 + p2 + p3 + p4 + ")";
    }
}

File: Python/백준/Silver/1992. 쿼드트리/쿼드트리.py
N = int(input())
arr = [input() for i in range(N)]


def check(x,y,N):
    start = arr[x][y]
    for i in range(x,x+N):
        for j in range(y,y+N):
            if start != arr[i][j]:
                return 1
    return 0

def aaa(x,y,N):
    if N==1:
        return str(arr[x][y])

    if check(x,y,N)==0:
        return str(arr[x][y])

    p1 = aaa(x,y,N//2)
    p2 = aaa(x,y+(N//2),N//2)
    p3 = aaa(x+(N//2),y,N//2)
    p4 = aaa(x+(N//2),y+(N//2),N//2)
    return '('+p1+p2+p3+p4+')'


print(aaa(0,0,N))

